### PR TITLE
Kernel+AK: Verify ref-counted objects' inheritance order, rearrange NVMeController inheritance

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

File: AK/RefCounted.h [hunk 1]
```diff
@@ -23,13 +23,19 @@ class RefCountedBase {
 
     ALWAYS_INLINE void ref() const
     {
+        // NOTE: If we can't find the matching signature, then RefCountedBase is not the first member
+        // of a Ref-counted object!
+        VERIFY(m_sig == 0x5245465054522020);
```
Comment (ADKaster) [inline AK/RefCounted.h:28]: This hack is not needed for RefCounted.

The following code won't even compile:

```c++
struct SomeBase {
    u64 m_base_0 = { 0xDEADBEEF0000CAFEULL };
    u64 m_base_1 = { 0xAA55AA55AA55AA55ULL };
};

struct Object3 : public SomeBase, public RefCounted<SomeBase>
{
    u64 m_object_0 = { 1122334455667788ULL };
};

TEST_CASE(not_first_base_class) {

    auto object = make_ref_counted<Object3>();

    EXPECT_EQ(object->m_base_0, 0xDEADBEEF0000CAFEULL);
    EXPECT_EQ(object->m_base_1, 0xAA55AA55AA55AA55ULL);
    EXPECT_EQ(object->ref_count(), 1U);
    EXPECT_EQ(object->m_object_0, 1122334455667788ULL);
}
```
with error
```
/home/andrew/serenity/Meta/Lagom/../../AK/RefCounted.h:59:37: error: static_cast from 'const AK::RefCounted<SomeBase> *' to 'const SomeBase *', which are not related by inheritance, is not allowed
        auto* that = const_cast<T*>(static_cast<T const*>(this));
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/andrew/serenity/Meta/Lagom/../../AK/NonnullRefPtr.h:33:14: note: in instantiation of member function 'AK::RefCounted<SomeBase>::unref' requested here
        ptr->unref();
             ^
/home/andrew/serenity/Meta/Lagom/../../AK/NonnullRefPtr.h:97:9: note: in instantiation of function template specialization 'AK::unref_if_not_null<Object3>' requested here
        unref_if_not_null(m_ptr);
        ^
```

We need to fix AtomicRefCounted to rely on more static casts to let the complier complain if it's wrong.

Comment (ADKaster) [reply AK/RefCounted.h:28]: Oh wait hold on, I think I messed up the CRTP

Comment (ADKaster) [reply AK/RefCounted.h:28]: Ok. now I'm very sure the bug is in LockRefPtr. Here's a complete example with the actual problem described in the commit message that works for RefPtr but not for LockRefPtr.

In order to compile it you need to 
```
#include <AK/AtomicRefCounted.h>
#include <Kernel/Library/LockRefPtr.h>
```
and remove the two `#ifdef KERNEL` blocks in AK/Forward.h

```c++
struct SomeBase {
    u64 m_base_0 = { 0xDEADBEEF0000CAFEULL };
    u64 m_base_1 = { 0xAA55AA55AA55AA55ULL };
};

struct OtherBaseRefCounted : public RefCounted<OtherBaseRefCounted>
{
    u64 m_ref_counted_0 = { 0xAABBCCDDEEFF00ULL };
};

struct Object3 : public SomeBase
    , public OtherBaseRefCounted {
    u64 m_object_0 = { 1122334455667788ULL };
};

TEST_CASE(not_first_base_class)
{

    auto object = make_ref_counted<Object3>();

    EXPECT_EQ(object->m_base_0, 0xDEADBEEF0000CAFEULL);
    EXPECT_EQ(object->m_base_1, 0xAA55AA55AA55AA55ULL);
    EXPECT_EQ(object->m_ref_counted_0, 0xAABBCCDDEEFF00ULL);
    EXPECT_EQ(object->ref_count(), 1U);
    EXPECT_EQ(object->m_object_0, 1122334455667788ULL);

    RefPtr<Object3> ref = object;

    EXPECT_EQ(object->m_base_0, 0xDEADBEEF0000CAFEULL);
    EXPECT_EQ(object->m_base_1, 0xAA55AA55AA55AA55ULL);
    EXPECT_EQ(object->m_ref_counted_0, 0xAABBCCDDEEFF00ULL);
    EXPECT_EQ(object->ref_count(), 2U);
    EXPECT_EQ(object->m_object_0, 1122334455667788ULL);
}

struct OtherBaseAtomicRefCounted : public AtomicRefCounted<OtherBaseAtomicRefCounted> {
    u64 m_ref_counted_0 = { 0xAABBCCDDEEFF00ULL };
};

struct Object4 : public SomeBase
    , public OtherBaseAtomicRefCounted {
    u64 m_object_0 = { 1122334455667788ULL };
};

TEST_CASE(not_first_base_class_atomic)
{
    NonnullLockRefPtr<Object4> object = adopt_lock_ref(*new Object4);

    EXPECT_EQ(object->m_base_0, 0xDEADBEEF0000CAFEULL);
    EXPECT_EQ(object->m_base_1, 0xAA55AA55AA55AA55ULL);
    EXPECT_EQ(object->m_ref_counted_0, 0xAABBCCDDEEFF00ULL);
    EXPECT_EQ(object->ref_count(), 1U);
    EXPECT_EQ(object->m_object_0, 1122334455667788ULL);

    LockRefPtr<Object4> ref = object;

    EXPECT_EQ(object->m_base_0, 0xDEADBEEF0000CAFEULL);
    EXPECT_EQ(object->m_base_1, 0xAA55AA55AA55AA55ULL);
    EXPECT_EQ(object->m_ref_counted_0, 0xAABBCCDDEEFF00ULL);
    EXPECT_EQ(object->ref_count(), 2U);
    EXPECT_EQ(object->m_object_0, 1122334455667788ULL);
}
```

**Comment by supercomputer7** [comment]
Let's close this because @Hendiadyoin1 has a better solution.
As for the `LockRefPtr` vs `RefPtr` problem, we can try to investigate the problem, but since they are really similar, I can't think of why this happens. Also, `LockRefPtr` should be phased out anyway.

**Pull request closed without merging**